File: ДЗ 2/ch-4-3-step-8.swift
struct MaxBinaryHeap {
    var rawData: [Int] = []

    mutating func insert(_ item: Int) {
        self.rawData.append(item)
        self.siftUp(at: self.rawData.endIndex - 1)
    }

    mutating func removeMax() -> Int? {
        guard self.rawData.count > 0 else {
            return nil
        }

        self.rawData.swapAt(0, self.rawData.endIndex - 1)
        let value = self.rawData.removeLast()

        if self.rawData.count > 0 {
            self.siftDown(at: 0)
        }

        return value
    }

    mutating private func siftUp(at index: Int) {
        guard index != 0 else {
            return
        }

        let parentIndex = self.parentIndexUnchecked(for: index)

        if self.value(at: index)! > self.value(at: parentIndex)! {
            self.rawData.swapAt(index, parentIndex)
            self.siftUp(at: parentIndex)
        }
    }

    mutating private func siftDown(at index: Int) {
        let value = self.value(at: index)!

        let children = self.childrenIndexesUnchecked(for: index)
        let leftValue = self.value(at: children.left)
        let rightValue = self.value(at: children.right)

        if leftValue != nil {
            if leftValue! > value {
                if rightValue != nil && rightValue! > leftValue! {
                    self.rawData.swapAt(index, children.right)
                    self.siftDown(at: children.right)
                } else {
                    self.rawData.swapAt(index, children.left)
                    self.siftDown(at: children.left)
                }
            } else if rightValue != nil && rightValue! > value {
                self.rawData.swapAt(index, children.right)
                self.siftDown(at: children.right)
            }
        }
    }

    func value(at rawIndex: Int) -> Int? {
        guard rawIndex < self.rawData.count else {
            return nil
        }

        return self.rawData[rawIndex]
    }

    private func parentIndexUnchecked(for childIndex: Int) -> Int {
        let (quotient, _) = MaxBinaryHeap.logicalIndex(for: childIndex).quotientAndRemainder(dividingBy: 2)
        return MaxBinaryHeap.rawIndex(for: quotient)
    }

    private func childrenIndexesUnchecked(for parentIndex: Int) -> (left: Int, right: Int) {
        let logicalParent = MaxBinaryHeap.logicalIndex(for: parentIndex)

        let logicalLeft = 2 * logicalParent
        let logicalRight = 2 * logicalParent + 1

        let left = MaxBinaryHeap.rawIndex(for: logicalLeft)
        let right = MaxBinaryHeap.rawIndex(for: logicalRight)

        return (left: left, right: right)
    }

    private static func logicalIndex(for rawIndex: Int) -> Int {
        return rawIndex + 1
    }

    private static func rawIndex(for logicalIndex: Int) -> Int {
        return logicalIndex - 1
    }
}

let operationCount = Int(readLine()!)!
var heap = MaxBinaryHeap()

for _ in 0..<operationCount {
    let line = readLine()!.split(separator: " ")

    if line[0] == "Insert" {
        let value = Int(line[1])!
        heap.insert(value)
    } else if line[0] == "ExtractMax" {
        let value = heap.removeMax()!
        print(value)
    }
}

File: ДЗ 2/ch-4-3-step-8.py
class MaxBinaryHeap:
    def __init__(self):
        self.rawData = []

    def insert(self, item):
        self.rawData.append(item)
        self.siftUp(len(self.rawData) - 1)

    def removeMax(self):
        if len(self.rawData) == 0:
            return None

        #self.rawData[0], self.rawData[-1] = self.rawData[-1], self.rawData[0]
        value = self.rawData.pop()

        if len(self.rawData) > 0:
            self.siftDown(0)

        return value

    def siftUp(self, index):
        if index == 0:
            return

        parentIndex = self.parentIndexUnchecked(index)

        if self.value(index) > self.value(parentIndex):
            self.rawData[index], self.rawData[parentIndex] = self.rawData[parentIndex], self.rawData[index]
            self.siftUp(parentIndex)

    def siftDown(self, index):
        value = self.value(index)

        children = self.childrenIndexesUnchecked(index)
        leftValue = self.value(children[0])
        rightValue = self.value(children[1])

        if leftValue is not None:
            if leftValue > value:
                if rightValue is not None and rightValue > leftValue:
                    self.rawData[index], self.rawData[children[1]] = self.rawData[children[1]], self.rawData[index]
                    self.siftDown(children[1])
                else:
                    self.rawData[index], self.rawData[children[0]] = self.rawData[children[0]], self.rawData[index]
                    self.siftDown(children[0])
            elif rightValue is not None and rightValue > value:
                self.rawData[index], self.rawData[children[1]] = self.rawData[children[1]], self.rawData[index]
                self.siftDown(children[1])

    def value(self, rawIndex):
        if rawIndex >= len(self.rawData):
            return None

        return self.rawData[rawIndex]

    @staticmethod
    def parentIndexUnchecked(childIndex):
        quotient = (childIndex - 1) // 2
        return MaxBinaryHeap.rawIndex(quotient)

    @staticmethod
    def childrenIndexesUnchecked(parentIndex):
        logicalParent = parentIndex + 1

        logicalLeft = 2 * logicalParent
        logicalRight = 2 * logicalParent + 1

        left = MaxBinaryHeap.rawIndex(logicalLeft - 1)
        right = MaxBinaryHeap.rawIndex(logicalRight - 1)

        return (left, right)

    @staticmethod
    def logicalIndex(rawIndex):
        return rawIndex + 1

    @staticmethod
    def rawIndex(logicalIndex):
        return logicalIndex - 1


operationCount = int(input())
heap = MaxBinaryHeap()

for i in range(operationCount):
    line = input().strip().split()

    if line[0] == "Insert":
        value = int(line[1])
        heap.insert(value)
    elif line[0] == "ExtractMax":
        value = heap.removeMax()
        if value is not None:
            print(value)
